0017-chart-namespace-overrides/README.md

```diff
@@ -0,0 +1,298 @@
+the ZEP is going to cover.
+-->
+
+<!--
+A table of contents helps reviewers quickly navigate the ZEP and highlights
+any additional information provided beyond the standard ZEP template.
+-->
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+- [Proposal](#proposal)
+  - [User Stories (Optional)](#user-stories-optional)
+    - [Story 1](#story-1)
+    - [Story 2](#story-2)
+  - [Risks and Mitigations](#risks-and-mitigations)
+- [Design Details](#design-details)
+  - [Test Plan](#test-plan)
+      - [Prerequisite testing updates](#prerequisite-testing-updates)
+      - [Unit tests](#unit-tests)
+      - [e2e tests](#e2e-tests)
+  - [Graduation Criteria](#graduation-criteria)
+  - [Upgrade / Downgrade Strategy](#upgrade--downgrade-strategy)
+  - [Version Skew Strategy](#version-skew-strategy)
+- [Implementation History](#implementation-history)
+- [Drawbacks](#drawbacks)
+- [Alternatives](#alternatives)
+- [Infrastructure Needed (Optional)](#infrastructure-needed-optional)
+<!-- /toc -->
+
+## Summary
+
+This ZEP proposes to enable a namespace override for charts similar to the namespace override [functionality available in UDS CLI](https://uds.defenseunicorns.com/reference/bundles/overrides/#namespace).  This would allow namespaces of charts within a Zarf package to be overridden so that multiples of the same Zarf package could be deployed to the same cluster under different namespaces (without needing to maintain variants of the same package).
+
+## Motivation
+
+Doing this allows more flexibility with certain Zarf packages where you may want to have multiples of them installed in the cluster with slightly different configurations (such as [GitLab Runners](https://github.com/defenseunicorns/uds-package-gitlab-runner)).  Right now the release namespace of any chart has to be hardcoded into the package and will be overwritten even if the chart allows namespace overrides for some objects within the chart.  The current behavior is also different from what Helm does by default which may not be what users of Zarf expect (Helm allows the use of the `namespace` flag on install to set the Chart's namespace without it needing to be baked into the Chart).
+
+### Goals
+
+- Provide a way for a Zarf package containing Helm Charts to be easily installed more than once with different configurations
+
+### Non-Goals
+
+- Move away from the declarative nature of Zarf packages
+
+## Proposal
+
+The proposed solution is to add a deploy configuration option that has a `map[string]string` referencing the names of charts within the Zarf Package correlated with their overridden namespaces.
+
+### User Stories (Optional)
+
+#### Story 1
+
+**As** Jacquline **I want** to be able to set namespace overrides **so that** I can install the same package with different configurations in different namespaces.
+
+##### Option 1 (Flatter Namespaces)
```

Comment:
I generally like option 1 the best, but wonder if "configurable-namespaces" should be a toggle in the package at create time? With some packages there's absolutely a desire/need to make the namespace flexible but others might struggle a lot with namespace flexibility (thinking of [uds-core](https://github.com/defenseunicorns/uds-core) for example). Having the package author explicitly allow or not allow configuring namespaces might be also align with keeping the declarative nature of zarf packages even more.

Tangentially related - if templates within a zarf package/helm chart needed to access the namespace (outside of the normal `metadata.namespace` usage) would this override be available as a zarf var or other value?

Reply:
A feature like `.metadata.configurable-namespaces` that defaults to true would make sense to solve that. For my understanding could you give a situation where allowing a user to configure their namespace might break a package?

Reply:
[UDS Core](https://github.com/defenseunicorns/uds-core) is probably the best example I can provide. When you have multiple helm charts in a single zarf package (in this case 10+) and each needs to be aware of the namespace that the other lives in so that things like networkpolicies are properly created (or even if there were svc -> svc connections across namespaces), making sure that all works would be pretty messy since `.Release.Namespace` is only available within the context of the single chart. Chart A couldn't auto-detect the namespace that Chart B is deployed in, so you'd have to add another mechanism to pass chart names for "siblings" around. I think it could over-complicate a package in a way that the package maintainer doesn't want to support 😄.

Reply:
Makes perfect sense, thanks for the explanation!

Reply:
In general I'm fine with option 1, but I gotta ask: what if a helm chart or other artifact wants to deploy into more than a single namespace? Will we warn about using `--namespace` in that case or at least discourage in the docs or maybe that's not even possible today?

Reply:
I believe this is now addressed in the new implementation - simple packages could take advantage of the global package namespace and if more namespaces were needed you could move to [Zarf Values](https://github.com/zarf-dev/proposals/pull/22) - if namespaces needed to be reused across components you could do that too.